[WIP] TELCODOCS-1540: Installing SNO clusters with the SiteConfig Operator (#78153)
openshift/openshift-docs: modules/cnf-understanding-templates.adoc

// Module included in the following assemblies:
// Epic CNF-9657 (TELCODOCS-1540) (4.16)
// * edge_computing/cnf-understanding-siteconfig-operator.adoc

:_mod-docs-content-type: CONCEPT
[id="cnf-understanding-templates_{context}"]
= Understanding templates

The {sco} supports the inclusion of Golang text templates in the object definitions, which allows you to dynamically generate installation manifests for target clusters.

> This reduces the need to create separate manifests for many clusters with similar configurations but with different values.

Author: WDYT?
> This way, you do not need to create separate manifests for many clusters that have similar configurations but with different values.

Additionally, the {sco} supports all link:https://masterminds.github.io/sprig/[sprig library functions] and provides two new functions that you can use while creating your custom manifests:

`toYaml`:: The `toYaml` function encodes an item into a YAML string. If the item cannot be converted to YAML the function returns an empty string.

`anyFieldDefined`:: The `anyFieldDefined` function checks if a specific field is defined. If defined, the {sco} renders the field as instructed.

.anyFieldDefined example from InfraEnv
[source,yaml]
----
[...]
{{ if (anyFieldDefined .Site.Proxy) }}
    proxy:
{{ .Site.Proxy | toYaml | indent 4 }}
{{ end }}
[...]
----

> There are two types of templates:

Reviewer: Would be great if we can flow into this section... sensing a bit of disconnect in the readability.

Author: I have moved this up a bit and rephrased the intro to the new functions to connect them

* Cluster-level templates that must reference only cluster-specific fields.
* Node-level templates that can reference both cluster- and node-specific fields.

The {sco} provides the following default, validated, immutable set of templates:

.Default set of templates
[cols=3*, width="95%", options="header"]
|====

|Installation method
|Template type
|Template content

.2+|Assisted Installer
a|Cluster-level templates +
(`ai-cluster-templates-v1.yaml`)
a|`ClusterDeployment` +
`InfraEnv` +
`KlusterletAddonConfig` +
`ManagedCluster` +
`AgentClusterInstall`

a|Node-level templates +
(`ai-node-templates-v1.yaml`)
a|`NMStateConfig` +
`BareMetalHost`

.2+|Image-based Install Operator
a|Cluster-level templates +
(`ibi-cluster-templates-v1.yaml`)
a|`ClusterDeployment` +
`KlusterletAddonConfig` +
`ManagedCluster` +

a|Node-level templates +
(`ibi-node-templates-v1.yaml`)
a|`ImageClusterInstall` +
`NetworkConfigMap` +
`BareMetalHost` +

|====

[id="cnf-special-template-variable_{context}"]
== Special templates variables

The {sco} provides a set of special template variables that you can use in your templates.

> `CurrentNode`:: This variable contains all the content for the current node being handled in templating.

Reviewer: ```suggestion
`CurrentNode`:: The SiteConfig operator explicitly controls the iteration of the node objects and thus exposes this variable as a means to access all the content for the current node being handled in templating.
```

`InstallConfigOverrides`:: This variable contains the merged `networkType`, `cpuPartitioningMode` and `installConfigOverrides` content.

> `ControlPlaneAgents`:: This variable is automatically derived from the `SiteConfig` node objects. It consists of the number of control plane agents.

Reviewer: ```suggestion
`ControlPlaneAgents`:: This variable is automatically derived from the `ClusterInstance` node objects. It consists of the number of control plane agents.
```

> `WorkerAgents`:: This variable is automatically derived from the `SiteConfig` node objects. It consists of the number of worker agents.

Reviewer: ```suggestion
`WorkerAgents`:: This variable is automatically derived from the `ClusterInstance` node objects. It consists of the number of worker agents.
```

You can create a custom templated field by capitalizing the field name in the text template.
For example, if you want to specify the `name` and `namespace` for your current node by using the `CurrentNode` special template variable, use the field names in the following form:

[source,yaml]
----
name: "{{ .SpecialVars.CurrentNode.HostName }}"

> namespace: "{{ .Site.ClusterName }}"

Reviewer: ```suggestion
namespace: "{{ .Spec.ClusterName }}"
```

----

[id="cnf-sync-wave-annotation_{context}"]
== Using the `cluster.open-cluster-management.io/sync-wave` annotation

> The `cluster.open-cluster-management.io/sync-wave` annotation controls in which order manifests are created, updated, or deleted.

Reviewer: ```suggestion
The `siteconfig.open-cluster-management.io/sync-wave` annotation controls in which order manifests are created, updated, and deleted.
```

The annotation takes an integer as a value and that integer constitutes as a wave.
You can add one or several manifests to a single wave.
If not specified for a resource, the annotation takes the default value of `0`.

When creating or updating resources, the {sco} reconciles the manifests in ascending order, while the {sco} deletes resources in descending order.

In the following example, if the {sco} creates or updated the manifests, the `AgentClusterInstall` and `ClusterDeployment` CRs are reconciled in the first wave, while `KlusterletAddonConfig` and `ManagedCluster` CRs are reconciled in the third wave.
If the {sco} deletes the resources, `KlusterletAddonConfig` and `ManagedCluster` CRs are the first to be deleted, while the `AgentClusterInstall` and `ClusterDeployment` CRs are the last.

> include::snippets/siteconfig-ai-cluster-templates-v1.adoc[]

Reviewer: Perhaps we can include a shorter synthetic example?

// Module included in the following assemblies:
// Epic CNF-9657 (TELCODOCS-1540) (4.16)
// * edge_computing/cnf-understanding-siteconfig-operator.adoc

:_mod-docs-content-type: CONCEPT
[id="cnf-understanding-templates_{context}"]
= Understanding templates

The {sco} supports the inclusion of Golang text templates in the object definitions, which allows you to dynamically generate installation manifests for target clusters.
This way, you do not need to create separate manifests for clusters that have similar configurations but with different values.

There are two types of templates:

* Cluster-level templates that must reference only cluster-specific fields.

> * Node-level templates that can reference both cluster- and node-specific fields.

Reviewer 2: typo "cluster-"

== Special templates variables

The {sco} provides a set of special template variables that you can use in your templates.

`CurrentNode`:: The {sco} explicitly controls the iteration of the node objects and exposes this variable to access all the content for the current node being handled in templating.
`InstallConfigOverrides`:: This variable contains the merged `networkType`, `cpuPartitioningMode` and `installConfigOverrides` content.
`ControlPlaneAgents`:: This variable is automatically derived from the `ClusterInstance` node objects. It consists of the number of control plane agents.
`WorkerAgents`:: This variable is automatically derived from the `ClusterInstance` node objects. It consists of the number of worker agents.

You can create a custom templated field by capitalizing the field name in the text template.
For example, if you want to specify the `name` and `namespace` for your current node by using the `CurrentNode` special template variable, use the field names in the following form:

[source,yaml]
----
name: "{{ .SpecialVars.CurrentNode.HostName }}"
namespace: "{{ .Spec.ClusterName }}"
----

[id="cnf-sync-wave-annotation_{context}"]

> == Using the `siteconfig.open-cluster-management.io/sync-wave` annotation

Reviewer 2: Should the heading here outline what this section is about, rather than how it's done. Perhaps something like, "Controlling the manifest order" or something like that
